File: 2192-all-ancestors-of-a-node-in-a-directed-acyclic-graph/2192-all-ancestors-of-a-node-in-a-directed-acyclic-graph.cpp
class Solution {
    void dfs(int start, int curr, vector<vector<int>>& adj, vector<vector<int>>& ans){
        for (int neighbor : adj[curr]){
            if (!ans[neighbor].empty() && ans[neighbor].back() == start) continue;
            ans[neighbor].push_back(start);
            dfs(start, neighbor, adj, ans);
        }
    }
    
public:
    vector<vector<int>> getAncestors(int n, vector<vector<int>>& edges) {
        vector adj(n, vector<int>()), ans(n, vector<int>());
        for(auto edge : edges){
            adj[edge[0]].push_back(edge[1]);
        }
        
        for (int node = 0; node < n; ++node)
            dfs(node, node, adj, ans);
        return ans;
    }
};

File: 2192-all-ancestors-of-a-node-in-a-directed-acyclic-graph/2192-all-ancestors-of-a-node-in-a-directed-acyclic-graph.py
class Solution:
    def getAncestors(self, n: int, edges: List[List[int]]) -> List[List[int]]:
        adj = [[] for _ in range(n)]
        indegrees = [0] * n
        
        for start, end in edges:
            adj[start].append(end)
            indegrees[end] += 1
        
        queue = deque()
        ans = [set() for _ in range(n)]
        for i in range(len(indegrees)):
            if indegrees[i] == 0:
                queue.append(i)
        
        while queue:
            curr = queue.popleft()
            
            for neighbor in adj[curr]:
                ans[neighbor].add(curr)
                ans[neighbor].update(ans[curr])
                indegrees[neighbor] -= 1
                if indegrees[neighbor] == 0:
                    queue.append(neighbor)
        
        ans = [(sorted(list(s))) for s in ans]
        return ans
        